### PR TITLE
Generate vshuff to interleave vectors.

### PR DESCRIPTION
**Pull request by pranavb-ca**
Added tests to both simd_op_check.cpp and test/hexagon/codegen. We
should deprecate test/hexagon/codegen only after all tests in that
directory have been ported to simd_op_check.cpp.

### CODE REVIEW

File: src/CodeGen_Hexagon.cpp [hunk 1]
```diff
@@ -2232,6 +2232,28 @@ void CodeGen_Hexagon::visit(const Call *op) {
...
+        Expr arg0 = op->args[0];
+        Expr arg1 = op->args[1];
+        if ((arg0.type() == arg1.type()) &&
+            op->type.element_of() == arg0.type().element_of()) {
+          internal_assert(op->type.lanes() == 2*arg0.type().lanes());
+          std::vector<Value *> Ops;
+          Ops.push_back(codegen(arg1));
+          Ops.push_back(codegen(arg0));
+          int scalar = -1 * op->type.bytes();
+          Value *scalar_value = codegen(IntImm::make(Int(32), scalar));
```
Comment (dsharletg) [inline src/CodeGen_Hexagon.cpp]: You should be able to just say codegen(scalar), no need to explicitly construct an IntImm. It's a bit easier to read IMO.

File: src/CodeGen_Hexagon.cpp [hunk 2]
```diff
@@ -2232,6 +2232,28 @@ void CodeGen_Hexagon::visit(const Call *op) {
...
+          Ops.push_back(scalar_value);
+          llvm::Function *F =
+            Intrinsic::getDeclaration(module.get(),
+                                      IPICK(Intrinsic::hexagon_V6_vshuffvdd));
+          Value * Call = CallLLVMIntrinsic(F, Ops);
```
Comment (dsharletg) [inline src/CodeGen_Hexagon.cpp]: How about:

CallLLVMIntrinsic(F, {codegen(arg1), codegen(arg0), codegen(scalar)})

?

It's a bit more expressive and quicker to see exactly what is happening here.

Comment (pranavb-ca) [inline src/CodeGen_Hexagon.cpp]: This may be a problem because this involves a non-const temporary. CallLLVMIntrinsic cannot take a second parameter that is 'const' unless it is re-written. Given how extensively it is used, I am averse to changing it for this one patch. I will try to make the above more concise yet expressive though.

File: src/CodeGen_Hexagon.cpp [hunk 3]
```diff
@@ -2232,6 +2232,24 @@ void CodeGen_Hexagon::visit(const Call *op) {
...
+    } else if (op->name == Call::interleave_vectors) {
+      if (isDblVector(op->type, native_vector_bits()) &&
+                      op->args.size() == 2) {
+        Expr arg0 = op->args[0];
+        Expr arg1 = op->args[1];
+        if ((arg0.type() == arg1.type()) &&
+            op->type.element_of() == arg0.type().element_of()) {
+          internal_assert(op->type.lanes() == 2*arg0.type().lanes());
+          int scalar = -1 * op->type.bytes();
+          std::vector<Value *> Ops = {codegen(arg1), codegen(arg0), codegen(scalar)};
```
Comment (dsharletg) [inline src/CodeGen_Hexagon.cpp]: This is better, thanks. But just out of curiosity, it looks like CallLLVMIntrinsic takes a non-const reference just so it can change some of the arguments prior to calling IRBuilder::CreateCall... could it just take a vector by value instead? Do callers actually use the modified arguments? The performance impact would probably be negligible, and I think it would be a saner method.

Comment (pranavb-ca) [inline src/CodeGen_Hexagon.cpp]: I agree with you. I am not too happy with the way we set up Ops before we call CallLLVIntrinsic. Changing it in the manner that you suggest will help things for I do not see any call site changing / using the arguments after the call to callLLVMIntrinsic. However, can we do that in a separate pull request?

Comment (dsharletg) [inline src/CodeGen_Hexagon.cpp]: Sure. I'll merge this now.
